### ISSUE TITLE
Unexpected behavior in generate with return by reference

### ISSUE THREAD

**Issue by vbeffara** [ericniebler/range-v3#807]
The following code gives a weird output:

```c++
#include <iostream>
#include <range/v3/all.hpp>
#include <vector>

const std::vector<int> p4{0, 1, 2, 3};

auto perms() {
    return ranges::view::generate([t = p4]() -> const std::vector<int> & {
        std::cout << "Returning vector of size " << t.size() << '\n';
        return t;
    });
}

int main(int /*argc*/, char ** /*argv*/) {
    auto ps = perms();
    for (const auto & p : ps | ranges::view::take(2)) std::cout << " -> Obtained vector of size " << p.size() << '\n';
    std::cout << "==========\n";
    for (const auto & p : perms() | ranges::view::take(2)) std::cout << " -> Obtained vector of size " << p.size() << '\n';
}
```

On both clang (OSX, trunk) and gcc (linux, 7.4) the output I obtain is this:
```
Returning vector of size 4
 -> Obtained vector of size 4
Returning vector of size 4
 -> Obtained vector of size 4
Returning vector of size 4
==========
Returning vector of size 4
 -> Obtained vector of size 0
Returning vector of size 4
 -> Obtained vector of size 4
Returning vector of size 4
```

Notice the `0` as the size of the first returned vector for the second version. I would not expect a difference between the two options here, is there something I am missing?

Related question, I would have expected the lambda passed to `generate` to be called 2 rather than 3 times with that code, are 3 calls the expected behavior?

**Comment by gnzlbg** [comment]
> Notice the 0 as the size of the first returned vector for the second version. I would not expect a difference between the two options here, is there something I am missing?

This looks like a bug to me.

> Related question, I would have expected the lambda passed to generate to be called 2 rather than 3 times with that code, are 3 calls the expected behavior?

The lambda is called once on the construction of the view, and the result is stored within the view. When an iterator is incremented, a new result is stored in the view. When you write:

```c++
int main(int /*argc*/, char ** /*argv*/) {
    auto ps = perms();
    std::cout << " -> Obtained vector of size " << ranges::begin(ps)->size() << '\n';
    std::cout << " -> Obtained vector of size " << ranges::begin(ps)->size() << '\n';
}
```

which prints:

```c++
Returning vector of size 4
 -> Obtained vector of size 4
 -> Obtained vector of size 4
```

dereferencing the iterators does not need to check whether there is a value in the view already, and if there isn't, create one, because the view always contains one value. 

When you do a `ps | view::take(2)` you are basically iterating over the range `[0, 2)`, with the begin iterator pointing to zero, and the end iterator pointing to 2. The range is empty when `begin == end`, so begin must be incremented 2 times to reach the end. A new value is generated every time the iterator is incremented independently of whether the result is read or not. So because one needs to iterate to one past the end to be able to do the `begin == end` comparison, the lambda is called a total of three times.

So I would say this is expected. Whether it's the right call, I don't know. This is an `InputRange`, so every time you dereference it the range is allowed to return a different value.  A way to get the lambda to only be called twice in your case would be to store the cached value in the iterators and to generate a next value on dereference only.

That would mean, however, that the code I posted above would call the lambda two times, returning a different value each time. and the iterators return two different values. And that code like this:

```c++
int main(int /*argc*/, char ** /*argv*/) {
    auto ps = perms();
    auto it = ranges::begin(ps);
    std::cout << " -> Obtained vector of size " << it>size() << '\n';
    std::cout << " -> Obtained vector of size " << it->size() << '\n';
}
```

would also call the lambda twice, returning two values each time.

**Comment by vbeffara** [comment]
OK, thanks a lot for the reply, I believe I see the point. I would have expected `take(2)` to advance only once but indeed the way it is now is closer to what `take_while` would do, and it makes sense. It does mean that `generate_n(f,2)` and `generate(f) | take(2)` have different semantics (the first one calls `f` twice, the second one 3 times), which is maybe surprising and could lead to issues if `f` has side effects, but I can live with that.

In any case, the undefined behavior remains.

**Comment by CaseyCarter** [comment]
Closing as a duplicate of #819, despite this issue being older, because #819 has a clearer title.

**Comment by vbeffara** [comment]
Wait, the initial bug I reported was about the undefined behavior, not about the number of times the function was called (that was incidental). As far as I am aware the main issue remains, and is not a duplicate of #819 - though it might well be fixed at the same time
